### ISSUE TITLE
Public API to get the value of inheritable properties

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit#2674]
AsyncVoidVerificationScope is a type that needs to give users a helpful failure rather than allowing itself to be used in more than one test run in parallel:

```cs
using NUnit.Framework;
using NUnit.Framework.Internal;

Assert.That(GetCurrentTestInheritableProperty("ParallelScope"), Is.EqualTo(ParallelScope.None),
    $"{nameof(AsyncVoidVerificationScope)} may only be used by tests that have [NonParallelizable] applied.");

private static object GetCurrentTestInheritableProperty(string name)
{
    for (var test = TestExecutionContext.CurrentContext.CurrentTest; test != null; test = test.Parent as Test)
        if (test.Properties.ContainsKey(name))
            return test.Properties.Get(name);

    return null;
}
```

Rather than grab into NUnit.Framework.Internal, it would be nice to have a helper method on ITest so that this is possible:

```cs
TestContext.CurrentContext.Test.GetInheritableProperty("ParallelScope")
```

It needs a clearer name, too.

Neither of these work inside the test method:
```cs
TestContext.CurrentContext.Test.Properties.Get("ParallelScope") // null
NUnit.Framework.Internal.TestExecutionContext.CurrentContext.ParallelScope // ParallelScope .Default
```

**Comment by CharliePoole** [comment]
We've discussed having inheritable properties in NUnit and (I thought) rejected the idea. 

I sometimes talk about Parallelizable and a few others as "inheritable" (i.e. using quotes) by which I mean "they look inherited to users but are not." I should probably stop doing that!!!

If we __actually__ start calling these things "inheritable" then the users will rightfully expect them to act in a way similar to inheritance. Could you explain what you mean in more words, without the label?

IMO this is especially important if we may ever reverse ourselves and actually have inheritable properties. My definition of such a property is that the same property and values appear on descendant tests, which is definitely not the case with ParallelizableAttribute.

I think what your example is looking for is whether the test is running on a parallel queue, which is easy to discover and could be exposed in the TestContext. If you wanted to know whether any other tests are __actually__ running in parallel, that would probably be impractical.

**Comment by jnm2** [comment]
I don't want to know what's _actually_ happening so much as I want to know what attributes are affecting it. In other words, there should still be _some_ NonParallelizable attribute applied even if the runners you are using at the time are not running anything in parallel. The `Properties` collections seem to be precisely what I need then.

Just to be clear: I'm not talking about class inheritance/C# polymorphism. I'm talking about inheritance in the same general sense in which fonts property settings are inherited from a form to its controls, or from an HTML element to the elements it contains. NUnit very clearly works via this type of inheritance when it decides which effective ParallelScope to use for each test. No matter what we term it, I need a configuration discovery API that follows this existing logic.

**Comment by CharliePoole** [comment]
What you state is how I understand inheritance in this context as well. But `ParallelScope` doesn't quite work that way. How we compute the "effective parallel scope" is pretty specific to `ParallelScope` and wouldn't apply to other attributes - it's not generalizable in my view. That's because `ParallelScope` includes flags that are used for two different purposes: to affect the current test and to affect the context in which child tests are run. In retrospect, I have regretted that design and I won't use it the next time, but it is what it is right now.

It seems to me that the most accurate description of what you need is something very specific like "Effective parallel scope." It could share the code that we actually use to decide whether to run things in parallel and thus would always be able to answer the question "If we were running in parallel, would this method be affected?"

One small issue: I don't think we actually set everything up to determine the effective scope if you are running using `--workers:0` Is that a problem for you? Or maybe a useful feature?

**Comment by CharliePoole** [comment]
BTW, I'm pushing on the terminology only because I'm pretty convinced we will eventually want __actual__ inheritable properties - ones that are visible when you query the test properties. I'd like to see the name reserved for them.

**Comment by jnm2** [comment]
That sounds good.

> I don't think we actually set everything up to determine the effective scope if you are running using --workers:0 Is that a problem for you? Or maybe a useful feature?

That's the problem I was trying to explain. I want to validate the existence and applicability of the configuration no matter what runner settings are used. `--workers:0` is a good example.

**Comment by CharliePoole** [comment]
Except that `--workers:0` is a pretty esoteric setting that basically says "Ignore anything we told you about running in parallel." If you think of it as a feature, it might be "We can determine if tests would have passed if they didn't use any parallelization." I considered removing it for a while, but it is useful for debugging nunit.

In any case, the computation of effective scope could be maintained even under that setting with some work. It would involve moving some code and changing the location of some #if statements.

**Comment by jnm2** [comment]
I'm worried about `--workers:0` just in principle. The real-world things I've hit are:

- the necessity of `#if !NCRUNCH [assembly: Parallelizable(ParallelScope.Children)]`
- the VS adapter not effectively running the tests in parallel

while the console does run these particular tests effectively in parallel.

**Comment by CharliePoole** [comment]
The first doesn't change anything fundamental about how the tests are run. The same code is used whether you have parallelizable attributes or not.

I have no experience with the second issue. The adapter doesn't run assemblies in parallel but we're talking here about what happens inside NUnit once it's invoked. That is unchanged.

**Comment by jnm2** [comment]
And I should add:

- running only one test at a time repeatedly while writing the tests

I especially want to detect the lack of [NonParallelizable] then.

**Comment by CharliePoole** [comment]
Sorry, I don't get that one. If you run a single test, it's still using the same queues and workers as when you run all the tests. Why is that different for you?

**Comment by jnm2** [comment]
@CharliePoole I don't anything about the queues and workers; only that `NUnit.Framework.Internal.TestExecutionContext.CurrentContext.ParallelScope` is consistently `Default` when I was expecting it to be `None`.

I want to be able to test this question: "Did you make sure in your source code that NUnit Framework will not choose to run this test in parallel with any other test, regardless of the runner configuration?"

**Comment by CharliePoole** [comment]
@jnm2 Changing the order of my answers from that of your questions for a clearer explanation.

When `Default` appears on the test itself, it means that no parallel scope was specified __on that test__. You get `None` when the test is explicitly marked as non-parallel. Internally, NUnit must distinguish between nothing being specified and something being specified. It's only when nothing is specified that NUnit looks at the `TestExecutionContext`.

When `Default` appears in the `TestExecutionContext`, it means that no parallelism will be used for descendants. Of course, NUnit looks at it before your test starts to run, but by the time you can see it would only refer to descendants of __your__ test. In the case of a test case, of course, there are no descendants, so it has no real meaning.

There is currently no public API available to answer your question. That's good, because we can design it right from the start rather than worry about compatibility. I'm asking questions because I want to nail down the user requirement we are trying to meet.

At this point, I understand your question except for "regardless of the runner configuration." Are you saying it should not be possible to write some kind of runner that forces your non-parallel tests to run in parallel? That doesn't strike me as achievable. Somebody could always instantiate the class, create two threads and run two methods. Can we restrict the condition to some set of existing runners?

One thing for sure: the framework has no settings that can be passed to it that will force non-parallel tests to run in parallel. It does have a setting that does the opposite - forcing parallel tests to run non-parallel.

**Comment by jnm2** [comment]
> At this point, I understand your question except for "regardless of the runner configuration." Are you saying it should not be possible to write some kind of runner that forces your non-parallel tests to run in parallel? That doesn't strike me as achievable. Somebody could always instantiate the class, create two threads and run two methods. Can we restrict the condition to some set of existing runners?

I'm stating it only for completeness. I avoided saying anything about what the runner might choose to do with parallelism. No matter what the runner does, what settings it might pass to the framework, etc, the framework itself should not choose parallelism for that test.

**Comment by CharliePoole** [comment]
Can't accept a catch-all because we can't catch everything. 😸 Many third parties create runners that manipulate our framework objects directly and your proviso sounded (to me) as if you were talking about those guys. We can, however, guarantee the answer for all runners that use the framework API.

Would giving you access to the selected Execution Strategy (Parallel, NonParallel, Direct) be helpful? That's an internal setting used by the parallel dispatcher. Parallel and NonParallel have the obvious meanings. Direct means to run on the same thread as the parent, whatever it is. Alternatively, we could create some new enumeration that publicly describes what you want to know.

**Comment by jnm2** [comment]
> Would giving you access to the selected Execution Strategy (Parallel, NonParallel, Direct) be helpful?

I don't know. Not if runner parameters might obscure the configuration defined in the code. I tend to favor a new API on `ITest` because it's less about the run and more about the definition of the test.

**Comment by CharliePoole** [comment]
Although runners can have lots of parameters, we only need be concerned about parameters accepted by the framework. We know that there is `Exactly(1)` such parameter that impacts parallel execution: `NumberOfTestWorkers`, which when set to zero suppresses parallelism.

My own view is that this is such an unusual setting that you probably don't need to worry about it. However, whoever does this could take it into account as an implementation detail in the PR.

I don't think this is something for ITest because whether a test runs in parallel is not simply determined by attributes on the test. The actual parallel/non-parallel execution is a property of the `WorkItem`, which we create at execution time rather than when we load the test. 

We have never exposed the `WorkItem` publicly to the user but have instead put the relevant info into the context. A `WorkItem`, as you know, is in a one-to-one relation with a particular context. So ultimately the context is not about the entire run but about a particular work item that executes a particular test. You would be asking the context whether the current execution is in a parallel environment, without regard to how we got there. I think that's what you want.

**Comment by jnm2** [comment]
Now that you parse it out so clearly, and I think about it, I've been leading you down the wrong path by focusing too much on the idea of a catch-all. I only really care about what is possible to specify in source code, and even further I want to enforce the presence of some source code (attribute) somewhere configuring `ParallelScope.None` for the current test and no other code (attribute) overruling it.

**Comment by CharliePoole** [comment]
So basically, you want to be assured that the framework will never run this method in parallel with anything else. That would be true if
 * It has a [NonParallelizable] attribute
 * It has no attribute but it's class has `[NonParallelizable]`
 * Neither method nor class has it attribute but some higher level containing suite has it.
 * No parent test suites all the way up to the top have [Parallelizable(ParallelScope.Child)]

By the time the test is running, we already know all that. The test will be running under control of a non-parallel worker. That's where we could get the information you want. You would only need more info if you needed to figure out __why__ the test is running under a particular type of worker, but I think that's more of a debugging exercise than anything else.

Does that make sense?

**Comment by jnm2** [comment]
So long as `NumberOfTestWorkers` would not give it a non-parallel worker, then yes.

But I don't want an API to query "Is this a non-parallel worker?", do I? Because that's not really the question. That's an implementation detail in the answer. The question is, "Does the source code force this test to always run non-parallel?"

**Comment by CharliePoole** [comment]
That's why I first suggested we base it on the internal execution strategy enum, since it tells how the test __might__ be run rather than how it __is__ being run.

Parallel - forced to run parallel
NonParallel - forced to run non-parallel
Direct - might be either

To be clear, I was not suggesting we expose that internal, just that we base something on it, since it is computed once per work item and saved.

**Comment by jnm2** [comment]
Okay, I'm with you. Now what should the API look like, since it sounds like you don't necessarily want it to be property-based?

**Comment by CharliePoole** [comment]
I'd make it a property of the current `TestContext`, since that indicates to me that it's always the same value being returned (for a particular context). I assume the result would be an enumeration. We should pick a name and values so they work together.

**Comment by jnm2** [comment]
:+1: What about `TestContext.ParallelScope` and the `ParallelScope` enum? Is that too leaky of an abstraction?

**Comment by CharliePoole** [comment]
I think `ParallelScope` has too many values that would not apply to what you are doing. As I stated in another issue, I would remove it if I could, since it actually implements two concepts. That said, going down all those values is instructive in terms of thinking about what applies to us.

Applicable to this issue:
 * Self - This test always runs in a parallel environment
 * None - This test always runs in a non-parallel environment
 * Default - Redefine as "could go either way"

Do not seem applicable
 * Child
 * Fixture
 * ItemMask
 * ContextMask

Thinking about names for values, I like Parallel, NonParallel and Parent (i.e. P or NP depending on the parent. I'd pull these from the internal execution strategy of Parallel, NonParallel and Direct.

Name of the property shouldn't have Scope, because we aren't talking about a scope but just about this single test. Perhaps keep the name ExecutionStrategy? Seems like that's what it is.

**Comment by jnm2** [comment]
If my check sees `ExecutionStrategy.Parent`, does that mean conclusively that `[NonParallelizable]` is or is not applied to the test or one of its containing tests?

**Comment by CharliePoole** [comment]
No. It could go either way. For your use case you'd need to see NonParallel

**Comment by CharliePoole** [comment]
No. It could go either way. For your use case you'd need to see NonParallel

**Comment by CharliePoole** [comment]
I would have to look at the code more closely to tell you under what circumstances that could appear.

**Comment by jnm2** [comment]
Is it possible for the check to see `Parent` instead of `NonParallel` when `[NonParallelizable]` is a applied to the class or assembly but not the method?

If  my check can't fully answer its question by making a single enum or bool == comparison, then ithe API is not really answering my question directly. It's still just an implementation detail of the answer, and I want NUnit to be responsible for the maintenance of such implementation details.

**Comment by CharliePoole** [comment]
In an earlier comment, I suggested we simply examine whether you are running on a parallel or non-parallel queue. It has to be one or the other. My understanding is that you wanted more info: whether the test itself had an attribute or if it came from higher up.

As you articulated the need further, I started suggesting more and more complicated solutions because your query seemed complicated, probably because I never understood the need completely.

SO, let's back up. Given that under the parallel dispatcher, you can only be executing from a parallel queue or a non-parallel queue, is there a reason you need to know more than that? IF there is no need, then we could simply have a boolean property `IsRunningParallel`.

Parallelism can get quite complex. You might want to know if you are running in parallel with other methods in the same class versus methods in different classes, etc. But if __all__ you need to know is whether anything else might run at the same time, it's pretty easy.

**Comment by jnm2** [comment]
Correct, I don't want to distinguish between the attribute being on the test itself or higher up. I just need to detect the effective (cascaded, inherited) configuration specified in the source code for the current test. The bool property `IsRunningParallel` sounds perfect.

**Comment by jnm2** [comment]
> We've discussed having inheritable properties in NUnit and (I thought) rejected the idea.

Continuing from https://github.com/nunit/nunit3-vs-adapter/issues/336#issuecomment-367006722, I haven't been able to find an existing issue discussing the pros and cons of inheritable properties. Should I start a new one?

**Comment by jnm2** [comment]
I'm thinking that generalizing this is not ultimately helpful.
I was really looking for the equivalent of an ITest method `ParallelScope GetEffectiveParallelScope()`.

@masaeedu was looking for something like an `ISet<string> GetAppliedCategories()` method.

Maybe these individual methods should be proposed instead, since the hierarchical semantics are so different?

**Comment by CharliePoole** [comment]
IMO, it's always better to generalize from specific cases that you have already implemented. As we have gotten more fearful of breaking changes, we have tended to try to get things "right the fist time" but I think that's usually a false hope.

I'd be OK to talk about "EffectiveParallelScope if you can define what the heck that means. 😸 I can think of two definitions right off the bat.

GetAppliedCategories seems like less of a good thing and has all sorts of complexities to it, at least if I correctly understand what it is.

Individual issues could at least spell out what we are talking about.

**Comment by stevenaw** [comment]
I think this has been added recently. TestContext now supports the following instance methods for the current test:
- `PropertyHierarchy()`
- `PropertyValues(string)`
- `AllPropertyValues(string)`
- `AllCategories()`
